californium-core/src/main/java/org/eclipse/californium/core/coap/Message.java

```diff
@@ -671,36 +614,45 @@ public EndpointContext getSourceContext() {
 	/**
 	 * Set destination endpoint context.
 	 * 
-	 * Multicast addresses are not supported.
+	 * Multicast addresses are only supported for {@link Request}s.
 	 * 
 	 * Provides a fluent API to chain setters.
 	 * 
 	 * @param peerContext destination endpoint context
 	 * @return this Message
 	 * @throws IllegalArgumentException if destination address is multicast
-	 *             address
+	 *             address, but message is no {@link Request}
 	 * @see #setRequestDestinationContext(EndpointContext)
+	 * @deprecated use {@link #setDestinationContext(EndpointContext)} of
+	 *             {@link Request}, {@link Response}, or {@link EmptyMessage}
+	 *             instead. The parameter validation depends too much on the
+	 *             context to use this function on the base-class.
 	 */
-	public Message setDestinationContext(EndpointContext peerContext) {
-		// requests calls setRequestDestinationContext instead
-		if (peerContext != null && peerContext.getPeerAddress().getAddress().isMulticastAddress()) {
-			throw new IllegalArgumentException("Multicast destination is only supported for request!");
-		}
-		this.destinationContext = peerContext;
-		return this;
-	}
+	@Deprecated
+	public abstract Message setDestinationContext(EndpointContext peerContext);
 
 	/**
 	 * Set destination endpoint context for requests.
 	 * Multicast addresses are supported.
 	 * 
 	 * @param peerContext destination endpoint context
 	 * @see #setDestinationContext(EndpointContext)
+	 * @deprecated use {@link #setInternalDestinationContext(EndpointContext)}
 	 */
+	@Deprecated
 	protected void setRequestDestinationContext(EndpointContext peerContext) {
 		this.destinationContext = peerContext;
 	}
 
+	/**
+	 * Set destination endpoint context.
+	 * 
+	 * @param peerContext destination endpoint context
+	 */
+	protected void setInternalDestinationContext(EndpointContext peerContext) {
```

Comment:
I do not get the benefice of adding this new method. Why not something like [this](https://github.com/eclipse/californium/commit/0df093515b368462a1101e0695de1ba2262acde7) ?

Reply:
Basically to deprecate the `Message.setDestinationContext(EndpointContext)` in order to make that `protected` with the next major version. Californium doesn't call it on the level of `Message`, it always has a `Request`, `Response` or `EmptyMessage` (just delete it from `Message`, it still compiles).